### DISCUSSION TITLE
[Feature Request] - Server Side Sourcemaps (production) for Node Stacktraces

### DISCUSSION THREAD

**Discussion by jasonwilliams** [vercel/next.js#20096] [Ideas]
NodeJS can now parse sourcemaps (for better stack traces) in v12.20.0 onwards. So it would be nice to use this feature (link below).
https://github.com/vercel/next.js/pull/13018 looks like it added support but just for the browser side? Does this generate sourcemaps for the server side too?

If not then this issue is a request for that.

* https://nodejs.medium.com/source-maps-in-node-js-482872b56116
* https://nodejs.org/dist/latest-v12.x/docs/api/cli.html#cli_enable_source_maps

**Comment by jasonwilliams** [comment]
@timneutkens is there any reason why we wouldn't want this? It seems like an easy addition and can be off by default. Its a shame some of us cant make use of Node's `--enable-source-maps` while this is blocked.

Something like the following would be useful
```js
module.exports = {
  experimental: {
    productionServerSourceMaps: true
  }
}
```

**Reply by Kluskey** [reply]
Any updates on this @jasonwilliams ?

**Reply by jasonwilliams** [reply]
Nope, I think @timneutkens is the person to ask

**Comment by jasonwilliams** [comment]
Any updates on this @leerob @timneutkens ?
I doubt this is difficult to implement, and is preventing people using stack traces

**Comment by EloB** [comment]
A feature that I would like is a way to easy create source maps for production. I know that already exists with `productionBrowserSourceMaps` but I would like to extend that setting so you can choose how to expose them. For security reasons I don't like to expose source maps but it would be really good if it was possible to make them available by having a header a cookie or something. I know that you can make this yourself but would be good if this was builtin.

A proposal would be to have something like this. This isn't a silver bullet if you put your content on a separate server but if the Next.js app is serving the static files.
```ts
type MaybePromise<T> = T | Promise<T>;

type ExposePublic = 'public';
type ExposeRequest = (req: Request) => MaybePromise<boolean>;
type ExposeHeaders = {
  type: 'cookie' | 'headers' | 'basic-auth',
  name?: string,
  match: string | (value: string) => MaybePromise<boolean>
};

type ProposalNextConfig = NextConfig & {
  sourceMaps: {
    production: boolean;
    exposeStrategy:
      ExposePublic |
      ExposeRequest |
      ExposeHeaders;
  }
};
```

**Comment by r-martins** [comment]
It seems this would resolve the issue of [debugging server-side functions](https://stackoverflow.com/questions/75529150/how-can-i-debug-generatestaticparams-next-js-in-phpstorm), right?

**Comment by tordans** [comment]
Does anyone have experience with https://www.highlight.io/blog/lw5-next-js-server-sourcemaps ? It claims to show a way to enable source maps on the server…

**Reply by DeeeeLAN** [reply]
This won't work for development, at least. Development doesn't allow changes to config.devtool

**Comment by apflieger** [comment]
So apparently we still don't have any way to have sourcemaps in production right?
Do people have nextjs app in prod even? I don't get it. Should we just don't care about errors in prod?

**Comment by cyounkins** [comment]
Run the server as `node --enable-source-maps server.js`

**Reply by apflieger** [reply]
We don't have access to the node command when using `next` binary.
This works in local but not on Vercel:

`package.json`
```
"scripts": {
    "start": "NODE_OPTIONS=\"--enable-source-maps\" next start",
  },
```
I suppose Vercel runs the command `next start` by itself. I couldn't fin much about this on the docs. I also try to put an environment variable `NODE_OPTIONS` but it has no effect.

**Reply by enyineer** [reply]
Did you configure webpack to include sourcemaps? This has to be done "manually" to enable sourcemaps for server side components together with `--enable-source-maps`.

In your `next.config.js`, add:
```
webpack(config, { isServer, dev }) {
  if (isServer && !dev) {
    config.devtool = 'source-map';  // emits .map files for server
  }
  return config;
}
```

I'm running self hosted in Docker (building my nextjs in standalone mode) but this works for me.

Make sure your output folder (`.next`) includes `.map` files where the chunks are generated by webpack.

PS: `--enable-source-maps` will only work in nodejs 20+ environments. If you are on nodejs 18 or earlier, try [source-map-support](https://www.npmjs.com/package/source-map-support).

**Reply by jjimenez22991** [reply]
@enyineer I'm currently using node 22 and this configuration but it's actually not working for API routes. Is this supposed to work with API routes too? The .map files are located inside .next/standalone/.next/server/chunks and I'm working with standalone output.

**Comment by bryanhpchiang** [comment]
+1

**Comment by timneutkens** [comment]
Update on this: This is the default with Turbopack builds. We've been able to implement sourcemaps in a way that does not have a negative effect on performance and memory usage. Currently available in alpha: https://nextjs.org/blog/next-15-3#turbopack-builds-alpha

**Reply by giulianovarriale** [reply]
Sorry, I couldn't make it work. How should I run the source maps with turbolinks?
I tried building the application with `next build --turbopack` and then `next start` and the stack trace was not mapped to my source code. I also tried with the `--enable-source-maps` flag and it didn't work.
